arbnode/sync_monitor.go

```diff
@@ -153,10 +159,34 @@ func (s *SyncMonitor) SafeBlockNumber(ctx context.Context) (uint64, error) {
 	if err != nil {
 		return 0, err
 	}
+	// If SafeBlockWaitForBlockValidator is true, we want to wait for the block validator to finish
+	if s.config.SafeBlockWaitForBlockValidator {
+		latestValidatedCount, err := s.getLatestValidatedCount()
+		if err != nil {
+			return 0, err
+		}
+		if msg > latestValidatedCount {
+			msg = latestValidatedCount
+		}
+	}
 	block := s.exec.MessageIndexToBlockNumber(msg - 1)
 	return block, nil
 }
 
+func (s *SyncMonitor) getLatestValidatedCount() (arbutil.MessageIndex, error) {
+	latestValidatedGS := s.txStreamer.validator.GetLastValidated()
```

Comment:
check if validator is nil first (if so - return an error)

Reply:
Done.